src/HttpAdapterFactory.php

```diff
@@ -81,4 +95,39 @@ public static function register($name, $class)
 
         self::$classes[$name] = $class;
     }
+
+    /**
+     * guesses the best matching adapter
+     *
+     * @param  array                $preferred
+     * @return HttpAdapterInterface
+     * @throws HttpAdapterException
+     */
+    public static function guess(array $preferred = array())
+    {
+        foreach ($preferred as $preference) {
+            if (self::capable($preference)) {
+                return self::create($preference);
+            }
+        }
+
+        foreach (self::$guessMap as $name => $clientClass) {
+            if (self::capable($name)) {
+                return self::create($name);
+            }
+        }
+
+        throw new HttpAdapterException('no suitable HTTP adapter found');
+    }
+
+    /**
+     * checks if its possible to create a specified adapter
+     *
+     * @param string $name
+     * @return bool
+     */
+    public static function capable($name)
+    {
+        return isset(self::$guessMap[$name]) && (class_exists(self::$guessMap[$name]) || function_exists(self::$guessMap[$name]) || ini_get(self::$guessMap[$name]));
```

Comment:
I would additionally add a check about the one available in self::$classes? Someone can register a custom classes and it should be taken in account too. IMO, if a name is registered in the classes and not in the guess map, it should be marked as capable and the extra check should not be trigerred.